docs/adr/0008-allowing-more-types-as-errors.md

```diff
@@ -0,0 +1,139 @@
+
+We create 2 types - the `interface` itself unchanged from how the struct would be generated now,
+and another exception "wrapper" for the type.
+It means you *catch* a slightly different name from the interface, but the non-error version of the interface is what you deal with.
+
+```
+try:
+    do_it()
+except ErrorDetailsException as e:
+    # There are 2 types here:
+    # `e` is an `ErrorDetailsException`
+    # `e.inner` (or any other name) is an `ErrorDetails` - does not derive from `Exception`
+    # We can make it look natural via delegation.
+    e.message()
+    # But also use the original interface
+    HandleErrorReportingProblem(e.inner)
+```
+
+To clarify, you could imagine generated Python code looking something like:
+```
+class ErrorDetails:
+    ...
+class ErrorDetailsException(Exception):
+    ...
+```
+### C) Create and original type with an error type nested within the original
+
+Very similar to the above - there are still 2 unique types created, but only one type is a top-level
+type - the error type is created inside the exception.
+
+The code catching exceptions might look something like:
+```
+try:
+    do_it()
+except ErrorDetails.Exception as e:
+    # same block above as for B
+```
+
+ie, there's no top-level `ErrorDetailsException` type, it's accessed via `ErrorDetails.Exception`
+
+To clarify, you could imagine generated Python code looking something like:
+```
+class ErrorDetails:
+    # snip regular interface.
+    class Exception(Exception):
+        # snip error specific code
+```
+
+## Decision Outcome
+
+Chosen option: "[option X]", because [justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force force | … | comes out best (see below)].
+
+## Pros and Cons of the Options
+
+### 1: Do nothing - don't support interfaces as errors.
+
+Con: Seems a bit negative :)
+
+### 2: Implement Option A above
+
+- Pro: no magic, there's exactly 1 type for the type in the UDL/macro
+- Con: That type gets "tainted" by being used as an error. Using the interface in other contexts is less natural.
+- Con: It might be tricky to use as an external type - all external consumers must use as an error, or none of them must.
+
+### 3: Implement Option B above:
+
+- Pro: a more "pure" system - the interface can be used naturally in contexts other than being an error.
+- Pro: It mirrors Rust features in that it "adds" a capability rather then transforms it.
+- Con: The wrapping might get in the way, and it's more complexity for a possibly non-existing use-case.
+
+### 4: Implement Option C above:
+
+- Pro and Cons roughly the same for Option B plus:
+- Pro: less global types
+- Pro: the relationship between the interface and error is cleaner
+- Con: Might feel less natural to foreign authors.
+
+### 5: Something else?
+
+* What would exposing a `dictionary` (ie, a struct with only data) mean? What other types might become errors, and how does that inform us?
```

Comment:
I think a dictionary makes sense.  It would work like a single enum variant, but it's base class would be `Exception` rather than the enum type.

Reply:
We'd then get to choose A, B or C here too. Would we want to force there to be exactly 1 record type which derives from `Exception` and not have a plain-old record available?

Something I also meant to mention above is a compatibilty concern: imagine you use a struct/record today as *not* an error. Trying to use them as an error would be a breaking change if it meant the name and layout changed *everywhere* it was used just because a single function in a single crate wanted to use it an an error.

Reply:
The renaming is indeed not great. My first instinct would be to not rename user types. 
Yes, that means you either don't have the `Exception` suffix in languages where it is an exception or you need to use the `Exception` suffix in Rust code, which also feels awkward.

Reply:
A dictionary translates into a class/struct in the foreign language, so it could inherit/implement the corresponding error/exception class/trait.
That still works in A and except from its inherit it doesn't change much, does it?

Reply:
We already have precedents for renaming error types - I think both Swift and Kotlin will rename "Error" to "Exception" and I *think* linters will complain otherwise.

(b) and (c) allows you to just append "Error" or "Exception" so you get this for free.

Reply:
> We'd then get to choose A, B or C here too. Would we want to force there to be exactly 1 record type which derives from Exception and not have a plain-old record available?

I think so

> Something I also meant to mention above is a compatibilty concern: imagine you use a struct/record today as not an error. Trying to use them as an error would be a breaking change if it meant the name and layout changed everywhere it was used just because a single function in a single crate wanted to use it an an error.

I would rephrase this as you simply couldn't repurpose the type as an error type, which doesn't seem so bad.  From the perspective of a foreign language consume, types are either exceptions or they're not and this restriction is completely natural.

This limitation would also be easy to overcome once we had more flexibility.  If I had a `Foo` type, I could always add a `FooError` type that was a dictionary with a single `foo` field.

Reply:
I changed this to reflect that even if we support other types, the same key considerations are also going to apply to them.